### ISSUE TITLE
Can IHP support multiple databases?

### ISSUE THREAD

**Issue by benjamin-thomas** [digitallyinduced/ihp#1469]
I'd like to test out IHP by working along side an existing app, backed by 200+ tables in postsgresql.

I'm thinking of implementing controller + views for existing tables and porting over, bit by bit, pieces of the application logic to the IHP side. Of course I wouldn't use the schema designer in this case.

If it goes well, I'd move more and more logic to IHP. I could then maybe even port over data to the IHP managed database to enable usage of the schema designer.

It seems difficult to achieve though, since IHP seems to want to own the whole stack so I'm requiring input, see if this is doable.

In short, I need my IHP dev web server to communicate with another postgres db on my dev network (over TCP).

Is it doable or not within the scope of the project?

Thanks :)

**Comment by mpscholten** [comment]
This should be easy doable if you don't use the Schema Designer + Data tool. I have not tried this but it might even work already if you manually pass the `DATABASE_URL` to the dev server:

```bash
export DATABASE_URL="postgresql://..."
./start
```

(You could even set this directly inside the `start` script)

Then IHP will still start it's own postgres process, but the dev server should use the database server specified inside `DATABASE_URL` instead of using the default built-in one.

If that doesn't work you can manually override the database url in `Config.hs` like this:

```haskell
config = do
    -- ...
    option (DatabaseUrl "postgres://...")
```

In both cases the app process will not use the built-in server anymore. The dev tools will continue to use the built-in one as they're not relying on the `DATABASE_URL` env var.

**Comment by benjamin-thomas** [comment]
Got ya, thanks. So if I understand correctly, IHP cannot deal with 2 databases for one app?

Also, assuming I can override the DATABASE_URL, I'd need to write something looking like the generated types. Any idea on how to deal with that?

I tried to import my schema, but that seemed like a no go. For a start, all my tables have INT primary keys so this doesn't fit the basic assumptions of the tooling.

**Comment by mpscholten** [comment]
> Got ya, thanks. So if I understand correctly, IHP cannot deal with 2 databases for one app?

By default no. It's generally doable if you manage the connections for the second database yourself (e.g. apps running on https://thin.dev/ run a single IHP app but connect to the right database based on the subdomain of a project).

> Also, assuming I can override the DATABASE_URL, I'd need to write something looking like the generated types. Any idea on how to deal with that?
> I tried to import my schema, but that seemed like a no go. For a start, all my tables have INT primary keys so this doesn't fit the basic assumptions of the tooling.

Importing the schema seems like the easiest way. INT PKs should work with most APIs (we've added support for this in the past, but very few people are using it).

**Comment by benjamin-thomas** [comment]
Ok thanks for the feedback :)
